aten/src/ATen/native/ReflectionPad.cpp

```diff
@@ -0,0 +1,228 @@
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+#include <tuple>
+
+namespace at {
+namespace native {
+
+namespace {
+
+template <typename scalar_t>
+static void reflection_pad1d_out_frame(
+    scalar_t *input_p, scalar_t *output_p,
+    int64_t nslices,
+    int64_t iwidth, int64_t owidth,
+    int64_t pad_l) {
+  int64_t i_start_x = std::max(int64_t(0), -pad_l);
+  int64_t o_start_x = std::max(int64_t(0), pad_l);
+
+  int64_t k, ip_x;
+#pragma omp parallel for private(k, ip_x)
+
+  for (k = 0; k < nslices; k++) {
+    for (int64_t j = 0; j < owidth; j++) {
+      if (j < pad_l) {
+        ip_x = pad_l * 2 - j;
+      } else if (j >= pad_l && j < iwidth + pad_l) {
+        ip_x = j;
+      } else {
+        ip_x = (iwidth + pad_l - 1) * 2 - j;
+      }
+      ip_x = ip_x - o_start_x + i_start_x;
+
+      scalar_t *dest_p = output_p + k*owidth + j;
+      scalar_t *src_p = input_p + k*iwidth + ip_x;
+      *dest_p = *src_p;
+    }
+  }
+}
+
+void reflection_pad1d_out_template(
+    Tensor& output, const Tensor& input_, IntList padding) {
+  int64_t dim_plane = 0;
+  int64_t dim_w = 1;
+  int64_t nbatch = 1;
+
+  AT_CHECK(input_.numel() > 0 &&
+    (input_.ndimension() == 2 || input_.ndimension() == 3), "non-empty 2D "
+    "or 3D (batch mode) tensor expected for input, but got: ", input_);
+
+  if (input_.ndimension() == 3) {
+    nbatch = input_.size(0);
+    dim_w++;
+    dim_plane++;
+  }
+
+  /* sizes */
+  auto pad_l = padding[0];
+  auto pad_r = padding[1];
+
+  int64_t nplane = input_.size(dim_plane);
+  int64_t input_w = input_.size(dim_w);
+  int64_t output_w  = input_w + pad_l + pad_r;
+
+  AT_CHECK(pad_l < input_w && pad_r < input_w, "Argument #4: Padding size "
+    "should be less than the corresponding input dimension, but got: padding (",
+    pad_l, ", ", pad_r, ") at dimension ", dim_w, " of input ", input_.sizes());
+
+  AT_CHECK(output_w >= 1 , 2,
+    "input (W: ", input_w, ")is too small. Calculated output W: ", output_w);
+
+  /* get contiguous input */
+  Tensor input = input_.contiguous();
+
+  /* resize output */
+  if (input.ndimension() == 2) {
+    output.resize_({nplane, output_w});
+
+    AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+      input.type(), "reflection_pad1d", [&] {
+        reflection_pad1d_out_frame<scalar_t>(
+          input.data<scalar_t>(), output.data<scalar_t>(),
+          nplane,
+          input_w, output_w,
+          pad_l);
+      }
+    );
+  } else {
+    output.resize_({nbatch, nplane, output_w});
+
+    int64_t p;
+#pragma omp parallel for private(p)
+    for (p = 0; p < nbatch; p++) {
+      AT_DISPATCH_FLOATING_TYPES_AND_HALF(
```

Comment:
you should put the dispatch outside the omp parallel loop, otherwise you are paying the cost each time (I'm surprised this doesn't show up in the benchmark).

Reply:
hmm, I tried to change it to the following, but it complains that:
> embedding a #pragma directive within macro arguments is not supported

```c++
    AT_DISPATCH_FLOATING_TYPES_AND_HALF(
      input.type(), "reflection_pad1d_backward", [&] {
        int64_t p;
#pragma omp parallel for private(p)
        for (p = 0; p < nbatch; p++) {
          reflection_pad1d_backward_out_frame(
            grad_input.data<scalar_t>() + p * nplane * input_w,
            grad_output.data<scalar_t>() + p * nplane * output_w,
            nplane,
            input_w, output_w,
            pad_l);
        }
      }
    );
```

Tried moving `#pragma` out as well, but got the following error.

> statement after '#pragma omp parallel for' must be a for loop

In terms of why it shows little difference in benchmark, I think maybe if we parallelize `p` perfectly (i.e., each or a small number of threads run on a different core), the overhead of `switch-case` multiple times in the macro is negligible. But I do agree with you that even if it only shows little impact on delay, it does wasting computational resources. 

In terms of avoiding the dispatch overhead in the loop, I can create a new `AT_DISPATCH_FLOATING_TYPES_AND_HALF_PARALLEL` macro and replace the existing one at appropriate locations. But since that will touch many files out of the scope of this PR, should we add that change here or make it a new PR? Or any other suggestions? Thanks! @gchanan

Reply:
I don't think you should write a `AT_DISPATCH_FLOATING_TYPES_AND_HALF_PARALLEL` macro.

You can probably just define a separate inline function to do the computation with the `pragma` inside and that will work and shouldn't affect performance (but you should check).  Also stackoverflow (i.e. https://stackoverflow.com/questions/3030099/pragma-in-define-macro) suggests _Pragma("...") should work but my initial attempts didn't work.  There's probably some super-smart macro you can write to make this all work, but it might be tricky.

Reply:
Thanks, will try.